### PR TITLE
[FLINK-24277][connector/kafka] And configuration for committing offset on checkpoint and disable it if group ID is not specified

### PR DESCRIPTION
**Pull request by PatrickRen**
## What is the purpose of the change
This pull request adds a configuration for KafkaSource for disabling offset commit on checkpoint, and add sanity checks on `group.id`

## Brief change log
- Add config for disabling offset commit on checkpoint
- Add sanity checks on `group.id`. If `group.id` is not specified, offset commit on checkpoint will be disabled
- Remove auto-generated group id in Kafka table source

## Verifying this change

This change added tests and can be verified as follows:

  - Added integration tests for using Kafka source without group ID
  - Added test validating KafkaSourceBuilder capturing corrupted configurations (enabled offset commit manually but missing group id)
  - Added test in Kafka table source for validating offset commit is disabled without group id

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java [hunk 1]
```diff
@@ -481,5 +489,24 @@ private void sanityCheck() {
                 "No subscribe mode is specified, "
                         + "should be one of topics, topic pattern and partition set.");
         checkNotNull(deserializationSchema, "Deserialization schema is required but not provided.");
+        // Check consumer group ID
+        checkState(
+                props.containsKey(ConsumerConfig.GROUP_ID_CONFIG) || !offsetCommitEnabledManually(),
+                String.format(
+                        "Property %s is required when offset commit is enabled",
+                        ConsumerConfig.GROUP_ID_CONFIG));
+    }
```
Comment (becketqin) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java:498]: Ideally, we should probably also throw exception if the committedOffsets initializer is used when group id was not specified. It is a little tricky because `ReaderHandledOffsetsInitializerader` is package private. 

One possible solution is to have an `OffsetInitializerValidator` interface. If an `OffsetInitializer` needs to be validated, it can implement that interface and the `KafkaSourceBuilder` will then validate the properties against that `OffsetInitializer`. At the beginning this interface could be private. We may consider exposing it to end users later if that turns out to be necessary.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java [hunk 2]
```diff
@@ -353,17 +353,19 @@ private void acquireAndSetStoppingOffsets(
...
-                        (tp, offsetAndMetadata) -> {
-                            Preconditions.checkNotNull(
-                                    offsetAndMetadata,
-                                    String.format(
-                                            "Partition %s should stop at committed offset. "
-                                                    + "But there is no committed offset of this partition for group %s",
-                                            tp, groupId));
-                            stoppingOffsets.put(tp, offsetAndMetadata.offset());
-                        });
+        if (!partitionsStoppingAtCommitted.isEmpty()) {
```
Comment (becketqin) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:356]: Why is this needed?

**Review by becketqin — COMMENTED** [review]
@PatrickRen Thanks for the patch. LGTM overall. Just one minor comment.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java (hunk 2 above)
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:356]: Thanks for the review @becketqin ! `consumer#committed` will check `group.id` in properties, but user can totally skip providing `group.id` if no partition stops at committed offset.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java (hunk 1 above)
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java:501]: Thanks for the suggestion @becketqin ! Implemented in 264a99c
